File: php_to_typescript/php_functions.js
//javascript to PHP
class PHP{
	static strtoupper(a){
		return a.toUpperCase();
	}
	static strlen(a){
		return a.length;
	}
	static isset(a){
		return (typeof a !== 'undefined');
	}
	static is_string(a){
		return (typeof a === 'string');
	}
	static is_array(a){
		return Array.isArray(a);
	}
	static is_callable(a){
		return typeof(a) === 'function';
	}
	static array_map(callback,array){
		return array.map(callback);
	}
	static strrev(a){
		return a.reverse();
	}
	static array_keys(a){
		return a.keys();
	}
	static array_reverse(a){
		return a.reverse();
	}
	static array_key_exists(key,arry){
		return key in array;
	}
	static array_pop(a){
		return a.pop();
	}
	static str_starts_with(haystack,needle){
		return haystack.startsWith(needle);
	}
	static str_ends_with(haystack,needle){
		return haystack.endsWith(needle);
	}
	static str_contains(haystack,needle){
		return haystack.includes(needle);
	}
	static array_reduce(array,callback){
		return array.reduce(callback);
	}
	static replace(search,subject,replace){
		return subject.replaceAll(search,replace);
	}
	static implode(separator,array){
		return array.join(separator);
	}
	static join(separator,array){
		return implode(separator,array);
	}
	static explode(separator,string){
		return string.split(separator);
	}
	static floatval(a){
		return parseInt(a);
	}
	static strval(a){
		return toString(a);
	}
	static substr(string,offset,length){
		return string.substring(offset,offset+length);
	}
	static strpos(string haystack, string needle){
		let to_return = haystack.indexOf(needle);
		if(to_return === -1) return false;
		else return to_return;
	}
	static sort(a){
		a.sort();
	}
	static array_push(a,b){
		return a.push(b);
	}
	static rand_array(a){
		return a[Math.floor(Math.random() * a.length)];
	}
	static shuffle(array){
		for (let i = array.length - 1; i > 0; i--) {
		return a.concat(b);
	}
			const j = Math.floor(Math.random() * (i + 1));
			[array[i], array[j]] = [array[j], array[i]];
		}
	}
	static random(){
		return Math.random();
	}
	static array_merge(a,b){
	static array_intersect(){
		var result = arguments[0];
		for(const i of arguments){
			result = i.filter(value => result.includes(value))
		}
		return result;
	}
	static array_filter(array, callback){
		return array.filter(callback);
	}
	static in_array(needle,haystack){
		return haystack.includes(needle);
	}
	static array_search(needle,haystack){
		return haystack.indexOf(needle);
	}
	static min(){
		return Math.min(arguments);
	}
	static max(){
		return Math.max(arguments);
	}
	static array_shift(a){
		return a.shift();
	}
	static array(...args){
		return args;
	}
	static call_user_func_array(callback,args){
		return callback.apply(this, args);
	}
	static json_encode(a){
		return JSON.stringify(a);
	}
	static json_decode(a){
		return JSON.parse(a);
	}
	static trim(a){
		return a.trim();
	}
	static array_unique(a){
		// from https://stackoverflow.com/questions/1960473/get-all-unique-values-in-a-javascript-array-remove-duplicates
		return = a.filter(function(value, index, self) {return self.indexOf(value) === index;});
	}
	static strrpos(haystack,needle){
		return haystack.lastIndexOf(needle);
	}
	static get_object_vars(a){
		return Object.values(a);
	}
	static array_splice(){
		let array = arguments[0];
		let offset = arguments[1];
		let length = arguments[2];
		let length = arguments[2];
	}
	static array_sum(a,b){
		// this answer is from
		return [1, 2, 3, 4].reduce((a, b) => a + b, 0);
	}
	static get_class(a){
		return.constructor.name;
	}
}

File: php_to_ruby/php_functions.rb
#ruby to php
def strtolower(a) return a.downcase end
def strtolower(a) return a.upcase end
def isset(a) return defined? a end
def is_array(a) return a.kind_of?(Array) end
def array_reverse(a) return a.reverse() end
def array_reduce(array,callback) return array.reduce(callback) end
def is_array(a) return a.kind_of?(Array) end

def strval(a)
# convert string to number
end

def echo(a)
puts(a)
end

def array_push(a,b)
return a.push!(b)
end

def array_pop(a)
return a.pop!()
end

def array_keys(a)
return a.keys
end

def array_key_exists(a)
return a.key? a
end

def shuffle(a)
a.shuffle!
end

def count(a)
	return a.length()
end

def array_rand(a)
	return a.sample
end

def array_merge(a,b)
	return a.concat(b)
end

def in_array(needle,haystack)
	return haystack.include? needle
end

def array_reverse(a)
	return a.reverse()
end

def array_search(needle,haystack)
	return haystack.index { |needle| ['b', 'c'].include?(needle) }
end

def array_intersect(a,b)
	return a & b
end

def array_shift(a)
	return a.shift()
end

def str_ends_with(haystack,needle)
	return haystack.end_with? needle
end

def str_starts_with(haystack,needle)
	return haystack.start_with? needle
end

# from https://rosettacode.org/wiki/Count_occurrences_of_a_substring#Ruby
def substr_count str, subStr
    return str.scan(subStr).length
end

def trim(a)
	return a.strip
end

def json_decode(a)
	return JSON.parse(a)
end

def strrpos(haystack,needle)
	return haystack.rindex(needle)
end

def array_values(a)
	return a.values
end

def class_name(a)
	return a.class.name
end

# use str.scan(subStr).length for substr_count
